### ISSUE TITLE
Time evolution of pressure

### ISSUE THREAD

**Issue by UmeshRoy** [openmm/openmm#3740]
Dear Developers,
         How can I get the time evolution of  pressure from simulation from openmm. I wan to check the average pressure of the system. 

Thank you in advance.

Umesh Roy

**Comment by jchodera** [comment]
OpenMM can't currently compute the atomic or molecular virial that is normally used to compute the pressure tensor (also discussed in https://github.com/openmm/openmm/issues/3647), so there is unfortunately no easy way to do that. The best solution we could offer right now is to suggest you might be able to do finite-difference estimates of the energy derivative with respect to box size, [from which the pressure could be obtained](https://aip.scitation.org/doi/full/10.1063/1.2363381). 

@peastman : I've labeled this question as a candidate for the [FAQ](https://github.com/openmm/openmm/wiki/Frequently-Asked-Questions).

**Comment by UmeshRoy** [comment]
Thank you. I appreciate your suggestions.

**Comment by kong1-731** [comment]
Is there a way to access the potential energy function directly in OpenMM so that we could easily use the finite-difference estimates of the energy derivative?

**Comment by peastman** [comment]
Use `context.getState(getEnergy=True).getPotentialEnergy()`.

**Comment by kong1-731** [comment]
Just to clarify, if we use context.getState to get the current potential energy, how would we be able to save the initial potential energy or the potential energy of the previous step in order to calculate the potential energy difference within a reporter? Would we have to calculate the pressure offline and not while we run the simulation?

Also, from the equation below in the paper referenced above ("The nature of the calculation of the pressure in molecular simulations of continuous models from volume perturbations"), it seems as though we would need to know if the volume is increasing or decreasing to calculate the change in configurational energy. Would you advise in our code to first calculate the difference in volume and depending on whether that is positive or negative proceeding to use either the forward or backward difference scheme? 

<img width="645" alt="Screen Shot 2022-08-19 at 3 43 50 PM" src="https://user-images.githubusercontent.com/70727844/185716287-71ba6cf7-cbdf-44cd-889a-0dfd9c6fe48d.png">

It is unclear to me how we could use the centered finite difference scheme in realtime when the volume fluctuations might not be of equal magnitude in the positive and negative directions, but I might be misunderstanding this. Thank you for any clarification.

<img width="648" alt="Screen Shot 2022-08-19 at 3 53 04 PM" src="https://user-images.githubusercontent.com/70727844/185716940-105f3031-26a8-45c5-9c08-56f837e43255.png">

**Comment by peastman** [comment]
I'd like to add computing pressure as a feature.  The complication is defining exactly what it means.  Pressure is a thermodynamic quantity, which means it's defined as an ensemble average, not a property of a single state.  You can define an "instantaneous pressure" from dE/dV, which can be calculated with a finite difference.  The problem is that the definition isn't unique.  Do you scale the box uniformly, or just along one axis?  Do you scale the position of each particle independently, or clusters connected by constraints, or whole molecules?  Each of these leads to a different "instantaneous pressure".

My suggestion is to put the function into the barostat.  Each barostat class already answers these questions.  For example, MonteCarloBarostat scales the box uniformly while MonteCarloAnisotropicBarostat scales each axis independently.  MonteCarloFlexibleBarostat has a `setScaleMoleculesAsRigid()` method to select how scaling should be done.  Each barostat class can add a `computeCurrentPressure()` method that computes it in a way that's consistent with how the barostat is doing pressure regulation.

This can work even if you want to run your simulation at constant volume.  Add a barostat to the system and call `setFrequency(0)` to disable it.

Does this sound like a good approach?

**Comment by kong1-731** [comment]
Yes, I think so. The ability to computeCurrentPressure() within the barostat and save that information seems to be consistent with the fact that the pressure computation may also be barostat-specific for GROMACS/ other MD languages.

**Comment by peastman** [comment]
I have the beginnings of an implementation of this in #4881.  Feedback on it would be very appreciated!
